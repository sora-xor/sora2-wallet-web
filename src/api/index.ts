import axiosInstance from 'axios'

export const BASE_URL = process.env.BASE_URL

const axios = axiosInstance.create()
axios.defaults.baseURL = BASE_URL

export default axios
