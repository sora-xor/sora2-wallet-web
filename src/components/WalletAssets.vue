<template>
  <div :class="computedClasses" v-loading="loading">
    <template v-if="assetsFiatAmount">
      <div class="total-fiat-values">
        <span class="total-fiat-values__title">{{ t('assets.totalAssetsValue') }}</span>
        <formatted-amount value-can-be-hidden is-fiat-value integer-only with-left-shift :value="assetsFiatAmount" />
      </div>
      <s-divider class="wallet-assets-item_divider" />
    </template>
    <s-scrollbar v-if="!!formattedAccountAssets.length" class="wallet-assets-scrollbar">
      <div class="wallet-assets-container">
        <template v-for="(asset, index) in formattedAccountAssets">
          <div class="wallet-assets-item s-flex" :key="asset.address">
            <i class="asset-logo" :class="nftIconClass(asset)" :style="getAssetIconStyles(asset.address)" />
            <div class="asset s-flex">
              <formatted-amount-with-fiat-value
                value-can-be-hidden
                value-class="asset-value"
                :value="getBalance(asset)"
                :font-size-rate="FontSizeRate.SMALL"
                :asset-symbol="asset.symbol"
                symbol-as-decimal
                :fiat-value="getFiatBalance(asset)"
                :fiat-font-size-rate="FontSizeRate.MEDIUM"
                :fiat-font-weight-rate="FontWeightRate.MEDIUM"
              >
                <div v-if="hasFrozenBalance(asset)" class="asset-value-locked p4">
                  <s-icon name="lock-16" size="12px" />
                  <span>{{ formatFrozenBalance(asset) }}</span>
                </div>
              </formatted-amount-with-fiat-value>
              <div class="asset-info">
                {{ asset.name || asset.symbol }}
                <s-tooltip :content="copyTooltip">
                  <span class="asset-id" @click="handleCopyAddress(asset.address)">
                    ({{ getFormattedAddress(asset) }})
                  </span>
                </s-tooltip>
              </div>
            </div>
            <s-button
              v-if="permissions.sendAssets"
              class="wallet-assets__button send"
              type="action"
              size="small"
              alternative
              :tooltip="t('assets.send')"
              :disabled="isZeroBalance(asset)"
              @click="handleAssetSend(asset)"
            >
              <s-icon name="finance-send-24" size="28" />
            </s-button>
            <s-button
              v-if="permissions.swapAssets"
              class="wallet-assets__button swap"
              type="action"
              size="small"
              alternative
              :tooltip="t('assets.swap')"
              @click="handleAssetSwap(asset)"
            >
              <s-icon name="arrows-swap-24" size="28" />
            </s-button>
            <s-button
              v-if="permissions.showAssetDetails"
              class="wallet-assets__button el-button--details"
              type="action"
              size="small"
              alternative
              :tooltip="t('assets.details')"
              @click="handleOpenAssetDetails(asset)"
            >
              <s-icon name="arrows-chevron-right-rounded-24" size="28" />
            </s-button>
          </div>
          <s-divider
            v-if="index !== formattedAccountAssets.length - 1"
            class="wallet-assets-item_divider"
            :key="`${asset.address}-divider`"
          />
        </template>
      </div>
    </s-scrollbar>
    <div v-else class="wallet-assets-empty">{{ t('assets.empty') }}</div>
    <s-button
      v-if="permissions.addAssets"
      class="wallet-assets-add s-typography-button--large"
      @click="handleOpenAddAsset"
    >
      {{ t('assets.add') }}
    </s-button>
  </div>
</template>

<script lang="ts">
import { Component, Mixins } from 'vue-property-decorator';
import { Getter, Action } from 'vuex-class';
import { FPNumber } from '@sora-substrate/util';
import type { AccountAsset } from '@sora-substrate/util/build/assets/types';
import { api } from '../api';

import FormattedAmountMixin from './mixins/FormattedAmountMixin';
import LoadingMixin from './mixins/LoadingMixin';
import CopyAddressMixin from './mixins/CopyAddressMixin';
import FormattedAmount from './FormattedAmount.vue';
import FormattedAmountWithFiatValue from './FormattedAmountWithFiatValue.vue';
import { RouteNames, HiddenValue } from '../consts';
import { getAssetIconStyles, formatAddress } from '../util';
import type { WalletPermissions } from '../consts';

@Component({
  components: {
    FormattedAmount,
    FormattedAmountWithFiatValue,
  },
})
export default class WalletAssets extends Mixins(LoadingMixin, FormattedAmountMixin, CopyAddressMixin) {
  @Getter accountAssets!: Array<AccountAsset>;
  @Getter withoutFiatAndApy!: boolean;
  @Getter permissions!: WalletPermissions;
  @Getter shouldBalanceBeHidden!: boolean;
  @Action navigate!: (options: { name: string; params?: object }) => Promise<void>;

  get computedClasses(): string {
    const baseClass = 'wallet-assets';
    const classes = [baseClass];

    if (this.assetsFiatAmount) {
      classes.push(`${baseClass}--fiat`);
    }

    return classes.concat('s-flex').join(' ');
  }

  get formattedAccountAssets(): Array<AccountAsset> {
    return this.accountAssets.filter((asset) => asset.balance && !Number.isNaN(+asset.balance.transferable));
  }

  get assetsFiatAmount(): Nullable<string> {
    if (!this.formattedAccountAssets || this.withoutFiatAndApy) {
      return null;
    }
    if (!this.formattedAccountAssets.length) {
      return '0';
    }
    const fiatAmount = this.formattedAccountAssets.reduce((sum: FPNumber, asset: AccountAsset) => {
      const price = this.getAssetFiatPrice(asset);
      return price
        ? sum.add(
            this.getFPNumberFromCodec(asset.balance.transferable, asset.decimals).mul(FPNumber.fromCodecValue(price))
          )
        : sum;
    }, new FPNumber(0));
    return fiatAmount ? fiatAmount.toLocaleString() : null;
  }

  nftIconClass(asset): string {
    return api.assets.isNft(asset) ? 'nft-asset' : '';
  }

  getFormattedAddress(asset: AccountAsset): string {
    return formatAddress(asset.address, 10);
  }

  getAssetIconStyles = getAssetIconStyles;

  getBalance(asset: AccountAsset): string {
    return `${this.formatCodecNumber(asset.balance.transferable, asset.decimals)}`;
  }

  isZeroBalance(asset: AccountAsset): boolean {
    return this.isCodecZero(asset.balance.transferable, asset.decimals);
  }

  hasFrozenBalance(asset: AccountAsset): boolean {
    return !this.isCodecZero(asset.balance.frozen, asset.decimals);
  }

  formatFrozenBalance(asset: AccountAsset): string {
    if (this.shouldBalanceBeHidden) {
      return HiddenValue;
    }
    return this.formatCodecNumber(asset.balance.frozen, asset.decimals);
  }

  handleAssetSwap(asset: AccountAsset): void {
    this.$emit('swap', asset);
  }

  handleAssetSend(asset: AccountAsset): void {
    this.navigate({ name: RouteNames.WalletSend, params: { asset } });
  }

  handleOpenAssetDetails(asset: AccountAsset): void {
    this.navigate({ name: RouteNames.WalletAssetDetails, params: { asset } });
  }

  handleOpenAddAsset(): void {
    this.navigate({ name: RouteNames.AddAsset });
  }
}
</script>

<style lang="scss">
.wallet-assets {
  &-scrollbar {
    @include scrollbar($basic-spacing-big);
  }
  .asset {
    .formatted-amount {
      display: block;
      width: 100%;
      line-height: var(--s-line-height-reset);
      &__container {
        justify-content: flex-start;
        text-align: left;
      }
      &--fiat-value {
        margin-top: $basic-spacing-mini;
      }
    }
    &-value {
      height: $basic-spacing-medium;
      font-size: var(--s-font-size-medium);
      font-weight: 800;
      letter-spacing: var(--s-letter-spacing-mini);
      line-height: var(--s-line-height-reset);
      &-locked {
        margin-left: $basic-spacing-extra-small;
      }
      .formatted-amount__decimal {
        font-weight: 600;
      }

      @include formatted-amount-tooltip;
    }
  }

  &__button.el-button.neumorphic.s-action:not(.s-primary).s-alternative {
    &:disabled {
      &,
      & > span > i {
        color: var(--s-color-base-background);
      }
    }
    &:not(:disabled) {
      &:hover,
      &:focus {
        color: var(--s-color-theme-accent-hover);
      }
      &:active,
      &.s-pressed {
        color: var(--s-color-theme-accent-pressed);
      }
    }
  }
}
</style>

<style scoped lang="scss">
@import '../styles/icons';

$wallet-assets-class: '.wallet-assets';
$wallet-assets-count: 5;

#{$wallet-assets-class} {
  flex-direction: column;
  margin-top: #{$basic-spacing-medium};

  &--fiat {
    #{$wallet-assets-class}-container {
      max-height: calc(#{$asset-item-height--fiat} * #{$wallet-assets-count} + 4px);
    }
    #{$wallet-assets-class}-item {
      height: $asset-item-height--fiat;
    }
  }

  &-container {
    max-height: calc((#{$asset-item-height} * #{$wallet-assets-count}) + 4px);
  }
  &-item {
    align-items: center;
    height: $asset-item-height;

    .asset {
      flex: 1;
      overflow-wrap: break-word;
      flex-direction: column;
      padding-right: var(--s-basic-spacing);
      padding-left: var(--s-basic-spacing);
      width: 30%;
      justify-content: center;
      &-value,
      &-info {
        line-height: var(--s-line-height-base);
      }
      &-info {
        margin-top: $basic-spacing-mini;
        @include hint-text(var(--s-line-height-reset));
        color: var(--s-color-base-content-primary);
      }
      &-id {
        outline: none;
        &:hover {
          text-decoration: underline;
          cursor: pointer;
        }
      }
      &-value-locked {
        display: inline-flex;
        align-items: center;
        background-color: var(--s-color-base-content-secondary);
        color: var(--s-color-base-on-accent);
        padding: 2px 7px;
        line-height: 1;
        border-radius: var(--s-border-radius-mini);
        max-width: 100%;
        > .s-icon-lock-16 {
          color: var(--s-color-base-on-accent);
        }
        > span {
          margin-left: #{$basic-spacing-mini};
          white-space: nowrap;
        }
      }
      &-converted {
        @include hint-text;
      }
    }
    .asset-logo {
      flex-shrink: 0;
      @include asset-logo-styles;
    }
    &_divider {
      margin: 0;
    }
  }
  &-add,
  &-empty {
    margin-top: #{$basic-spacing-medium};
  }
  &-empty {
    text-align: center;
    @include hint-text;
  }
  &__button {
    & + & {
      margin-left: 0;
    }
  }

  .total-fiat-values {
    display: flex;
    align-items: baseline;
    justify-content: center;
    flex-wrap: wrap;
    padding-top: #{$basic-spacing-extra-small};
    padding-bottom: #{$basic-spacing-extra-small};
    text-align: center;
    &__title {
      text-transform: uppercase;
      padding-right: #{$basic-spacing-extra-mini};
      white-space: nowrap;
      font-weight: 400;
      letter-spacing: var(--s-letter-spacing-small);
    }
    .formatted-amount--fiat-value {
      display: block;
      font-size: var(--s-font-size-medium);
      font-weight: 600;
    }
  }
}

.nft-asset {
  background-image: none !important;
  background-color: #f4f0f1 !important;
  position: relative;
  font-style: unset;

  &::before {
    content: 'NFT';
    position: absolute;
    font-weight: 800;
    top: 28%;
    left: 14%;
    color: var(--s-color-base-content-tertiary);
  }
}
</style>
