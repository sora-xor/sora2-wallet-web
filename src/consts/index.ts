import type { FPNumber, Operation } from '@sora-substrate/util';

export const HiddenValue = '******';

export const BLOCK_PRODUCE_TIME = 6_000;
export const MAX_ALERTS_NUMBER = 5;

export enum AppWallet {
  GoogleDrive = 'google-drive',
  FearlessWallet = 'fearless-wallet',
  PolkadotJS = 'polkadot-js',
  SubwalletJS = 'subwallet-js',
  TalismanJS = 'talisman',
}

export enum AccountActionTypes {
  Rename = 'rename',
  Export = 'export',
  Logout = 'logout',
  Delete = 'delete',
}

export enum RouteNames {
  InternalConnection = 'InternalConnection',
  WalletConnection = 'WalletConnection',
  WalletSend = 'WalletSend',
  Wallet = 'Wallet',
  WalletAssetDetails = 'WalletAssetDetails',
  CreateToken = 'CreateToken',
  ReceiveToken = 'ReceiveToken',
  AddAsset = 'AddAsset',
  SelectAsset = 'SelectAsset',
}

export enum WalletTabs {
  Assets = 'WalletAssets',
  History = 'WalletHistory',
}

export enum TokenTabs {
  Token = 'CreateSimpleToken',
  NonFungibleToken = 'CreateNftToken',
}

export enum AddAssetTabs {
  Token = 'AddAssetToken',
  NFT = 'AddAssetNFT',
}

export enum WalletFilteringOptions {
  All = 'All',
  Currencies = 'Currencies',
  NFT = 'NFT',
}

export enum SoraNetwork {
  Dev = 'Dev',
  Test = 'Test',
  Stage = 'Stage',
  Prod = 'Prod',
}

export enum HashType {
  ID = 'id',
  Block = 'block',
  Account = 'account',
  EthAccount = 'ethAccount',
  EthTransaction = 'ethTransaction',
}

export enum ExplorerType {
  Sorascan = 'sorascan',
  Subscan = 'subscan',
  Polkadot = 'polkadot',
}

export type ExplorerLink = {
  type: ExplorerType;
  value: string;
};

export enum Step {
  CreateSimpleToken = 'CreateSimpleToken',
  ConfirmSimpleToken = 'ConfirmSimpleToken',
  CreateNftToken = 'CreateNftToken',
  ConfirmNftToken = 'ConfirmNftToken',
  Warn = 'Warn',
}

export enum LoginStep {
  Welcome = 'Welcome',
  Import = 'Import',
  ImportCredentials = 'Import/Credentials',
  ImportExternalExtensionList = 'ImportExternal/ExtensionList',
  ImportExternalAcccountList = 'ImportExternal/AccountList',
  SeedPhrase = 'Create/SeedPhrase',
  ConfirmSeedPhrase = 'Create/ConfirmSeedPhrase',
  CreateCredentials = 'Create/Credentials',
  AccountList = 'AccountList',
}

export const AccountImportInternalFlow = [LoginStep.Import, LoginStep.ImportCredentials];
export const AccountImportExternalFlow = [LoginStep.ImportExternalExtensionList, LoginStep.ImportExternalAcccountList];
export const AccountCreateFlow = [LoginStep.SeedPhrase, LoginStep.ConfirmSeedPhrase, LoginStep.CreateCredentials];

export interface WalletPermissions {
  addAssets?: boolean;
  addLiquidity?: boolean;
  bridgeAssets?: boolean;
  createAssets?: boolean;
  sendAssets?: boolean;
  showAssetDetails?: boolean;
  swapAssets?: boolean;
}

export type EthBridgeUpdateHistory = (updateWalletHistory: VoidFunction) => Promise<void>;

export interface WalletAssetFilters {
  option: WalletFilteringOptions;
  verifiedOnly: boolean;
  zeroBalance: boolean;
}

export interface WalletInitOptions {
  withoutStore?: boolean;
  permissions?: WalletPermissions;
  updateEthBridgeHistory?: EthBridgeUpdateHistory;
}

export interface NetworkFeeWarningOptions {
  type: Operation;
  isXor?: boolean;
  amount?: FPNumber;
}

export enum FontSizeRate {
  SMALL = 'small',
  MEDIUM = 'medium',
  NORMAL = 'normal',
}

export enum FontWeightRate {
  SMALL = 'small',
  MEDIUM = 'medium',
  NORMAL = 'normal',
}

export enum LogoSize {
  MINI = 'mini',
  SMALL = 'small',
  MEDIUM = 'medium',
  BIG = 'big',
  BIGGER = 'bigger',
  LARGE = 'large',
}

export enum PaginationButton {
  First = 'first',
  Prev = 'prev',
  Next = 'next',
  Last = 'last',
}

// states for Ethereum Bridge transaction
export enum ETH_BRIDGE_STATES {
  INITIAL = 'INITIAL',
  SORA_SUBMITTED = 'SORA_SUBMITTED',
  SORA_PENDING = 'SORA_PENDING',
  SORA_REJECTED = 'SORA_REJECTED',
  SORA_COMMITED = 'SORA_COMMITED',
  EVM_SUBMITTED = 'EVM_SUBMITTED',
  EVM_PENDING = 'EVM_PENDING',
  EVM_REJECTED = 'EVM_REJECTED',
  EVM_COMMITED = 'EVM_COMMITED',
}

export const ObjectInit = () => null;

/**
 * DO NOT IMPORT THIS CONST!
 *
 * Contains wallet-specific words which shouldn't be translated.
 * It's used in TranslationMixin of SORA Wallet project and it's extended in Polkaswap TranslationMixin.
 */
export const TranslationConsts = {
  Ethereum: 'Ethereum',
  Etherscan: 'Etherscan',
  Hashi: 'HASHI',
  PolkadotJs: 'Polkadot{.js}',
  Sora: 'SORA',
  TBC: 'TBC',
  XYK: 'XYK',
  NFT: 'NFT',
  Polkadot: 'Polkadot',
  SORAScan: 'SORAScan',
  Subscan: 'Subscan',
  QR: 'QR', // QR code
  IPFS: 'IPFS', // Inter-planetary file system
  // SORA networks
  soraNetwork: {
    Dev: 'SORA Devnet',
    Test: 'SORA Testnet (private)',
    Stage: 'SORA Testnet',
    Prod: 'SORA Mainnet',
  },
  JSON: '.json',
} as const;
