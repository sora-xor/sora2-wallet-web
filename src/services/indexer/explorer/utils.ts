import { formatStringNumber } from '../../../util';

import type { SubqueryAssetEntity } from '../subquery/types';
import type { SubsquidAssetEntity } from '../subsquid/types';
import type { FiatPriceObject, UpdatesStream, PoolApyObject } from '../types';

export function parseAssetFiatPrice(entity: SubsquidAssetEntity | SubqueryAssetEntity): FiatPriceObject {
  const acc = {};
  const id = entity.id;
  const priceFPNumber = formatStringNumber(entity.priceUSD);
  const isPriceFinity = priceFPNumber.isFinity();
  if (isPriceFinity) {
    acc[id] = priceFPNumber.toCodecString();
  }
  return acc;
}

export function parsePriceStreamUpdate(entity: UpdatesStream): Nullable<FiatPriceObject> {
  if (!entity?.data) return null;

  const data = JSON.parse(entity.data);

  return Object.entries(data).reduce((acc, [id, price]) => {
    const priceFPNumber = formatStringNumber(price as string);
    const isPriceFinity = priceFPNumber.isFinity();
    if (isPriceFinity) {
      acc[id] = priceFPNumber.toCodecString();
    }
    return acc;
  }, {});
}

export function parseApyStreamUpdate(entity: UpdatesStream): PoolApyObject {
  const data = entity?.data ? JSON.parse(entity.data) : {};

  return Object.entries(data).reduce((acc, [id, apy]) => {
    const strategicBonusApyFPNumber = formatStringNumber(apy as string);
    const isStrategicBonusApyFinity = strategicBonusApyFPNumber.isFinity();
    if (isStrategicBonusApyFinity) {
      acc[id] = strategicBonusApyFPNumber.toCodecString();
    }
    return acc;
  }, {});
}
