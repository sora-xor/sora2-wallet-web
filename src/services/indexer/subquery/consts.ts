export const SubstrateEvents = {
  CurrenciesTransferred: {
    method: 'Transferred',
    section: 'currencies',
  },
};
