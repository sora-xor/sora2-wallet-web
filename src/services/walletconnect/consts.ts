export const WALLET_CONNECT_PROJECT_ID = 'feeab08b50e0d407f4eb875d69e162e8';
