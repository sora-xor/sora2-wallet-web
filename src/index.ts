import Vue from 'vue'
import { Store } from 'vuex'
import debounce from 'lodash/fp/debounce'

import installWalletPlugins from './plugins'
// import './styles' We don't need it for now

import SoraNeoWallet from './SoraNeoWallet.vue'
import WalletAvatar from './components/WalletAvatar.vue'
import InfoLine from './components/InfoLine.vue'
import FormattedAmount from './components/FormattedAmount.vue'
import FormattedAmountWithFiatValue from './components/FormattedAmountWithFiatValue.vue'
import NumberFormatterMixin from './components/mixins/NumberFormatterMixin'
import FormattedAmountMixin from './components/mixins/FormattedAmountMixin'
import { Components, Modules, FontSizeRate, FontWeightRate } from './types/common'
import en from './lang/en'
import internalStore, { modules } from './store' // `internalStore` is required for local usage
import { storage } from './util/storage'
import { api, connection } from './api'
import { delay, getExplorerLink } from './util'
import * as WALLET_CONSTS from './consts'

let store: Store<unknown>
let isWalletLoaded = false
let unsubscribeStoreFromStorage: Function

const subscribeStoreToStorageUpdates = store => {
  if (typeof unsubscribeStoreFromStorage === 'function') {
    unsubscribeStoreFromStorage()
  }

  const syncWithStorageHandler = debounce(100)(() => store.dispatch('syncWithStorage'))

  window.addEventListener('storage', syncWithStorageHandler)

  unsubscribeStoreFromStorage = () => window.removeEventListener('storage', syncWithStorageHandler)
}

const components = [
  { component: SoraNeoWallet, name: Components.SoraNeoWallet }
]

const SoraNeoWalletElements = {
  install (vue: typeof Vue, options: any): void {
    if (!options.store) {
      throw new Error('Please provide vuex store.')
    }
    Object.values(Modules).forEach(molude => {
      options.store.registerModule(molude, modules[molude])
    })
    store = options.store
    installWalletPlugins(vue, store)
    components.forEach(el => vue.component(el.name, el.component))
  }
}

if (typeof window !== 'undefined' && window.Vue) {
  window.Vue.use(SoraNeoWalletElements, {})
}

async function initWallet ({
  withoutStore = false,
  permissions
}: WALLET_CONSTS.WalletInitOptions = {}): Promise<void> {
  isWalletLoaded = false
  if (!withoutStore && !store) {
    await delay()
    return await initWallet({ withoutStore, permissions })
  } else {
    if (withoutStore) {
      store = internalStore
    }
    if (connection.loading) {
      await delay()
      return await initWallet({ withoutStore, permissions })
    }
    if (!connection.api) {
      await connection.open()
      console.info('Connected to blockchain', connection.endpoint)
    }
    if (permissions) {
      store.dispatch('setPermissions', permissions)
    }
    await api.initialize()
    await store.dispatch('getWhitelist')
    await store.dispatch('subscribeOnFiatPriceAndApyObjectUpdates')
    await store.dispatch('checkSigner')
    await store.dispatch('syncWithStorage')
    await store.dispatch('getAccountAssets')
    await store.dispatch('updateAccountAssets')
    subscribeStoreToStorageUpdates(store)
    isWalletLoaded = true
  }
}

export {
  initWallet,
  isWalletLoaded,
  en,
  api,
  connection,
  storage,
  getExplorerLink,
  SoraNeoWallet,
  WALLET_CONSTS,
  WalletAvatar,
  InfoLine,
  FormattedAmount,
  FormattedAmountWithFiatValue,
  NumberFormatterMixin,
  FormattedAmountMixin,
  FontSizeRate,
  FontWeightRate
}
export default SoraNeoWalletElements
