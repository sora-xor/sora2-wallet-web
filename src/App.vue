<template>
  <s-design-system-provider :value="libraryDesignSystem" id="app" class="app-wrapper">
    <s-button class="theme-switch" @click="changeTheme">{{ libraryTheme }} theme</s-button>
    <div class="wallet-wrapper s-flex">
      <sora-wallet />
    </div>
    <s-button class="hide-balance-switch" @click="toggleHideBalance">
      {{ shouldBalanceBeHidden ? 'hidden' : 'visible' }} balances
    </s-button>
  </s-design-system-provider>
</template>

<script lang="ts">
// This file is only for local usage
import { Component, Mixins, Watch } from 'vue-property-decorator';
import { Action, Getter } from 'vuex-class';

import { FPNumber, History } from '@sora-substrate/util';
import { switchTheme } from '@soramitsu/soramitsu-js-ui/lib/utils';
import type Theme from '@soramitsu/soramitsu-js-ui/lib/types/Theme';
import type DesignSystem from '@soramitsu/soramitsu-js-ui/lib/types/DesignSystem';

import TransactionMixin from './components/mixins/TransactionMixin';
import { initWallet } from './index';
import SoraWallet from './SoraWallet.vue';
import { SoraNetwork, NFT_STORAGE_API_KEY } from './consts';
import { ApiKeysObject } from './types/common';

@Component({
  components: { SoraWallet },
})
export default class App extends Mixins(TransactionMixin) {
  @Getter shouldBalanceBeHidden!: boolean;
  @Action toggleHideBalance!: AsyncVoidFn;

  @Getter libraryDesignSystem!: DesignSystem;
  @Getter libraryTheme!: Theme;
  @Getter firstReadyTransaction!: Nullable<History>;

  @Action trackActiveTransactions!: AsyncVoidFn;
  @Action resetActiveTransactions!: AsyncVoidFn;
  @Action resetAccountAssetsSubscription!: AsyncVoidFn;
  @Action resetRuntimeVersionSubscription!: AsyncVoidFn;
  @Action resetFiatPriceAndApySubscription!: AsyncVoidFn;
  @Action setApiKeys!: (apiKeys: ApiKeysObject) => AsyncVoidFn;
  @Action setSoraNetwork!: (network: SoraNetwork) => Promise<void>;

  async created(): Promise<void> {
    await this.setApiKeys({ nftStorage: NFT_STORAGE_API_KEY });
    await this.setSoraNetwork(SoraNetwork.Dev);
    await initWallet({ withoutStore: true, whiteListOverApi: true }); // We don't need storage for local development
    this.trackActiveTransactions();
    const localeLanguage = navigator.language;
    FPNumber.DELIMITERS_CONFIG.thousand = Number(1000).toLocaleString(localeLanguage).substring(1, 2);
    FPNumber.DELIMITERS_CONFIG.decimal = Number(1.1).toLocaleString(localeLanguage).substring(1, 2);
  }

  @Watch('firstReadyTransaction', { deep: true })
  private handleNotifyAboutTransaction(value: History): void {
    this.handleChangeTransaction(value);
  }

  beforeDestroy(): void {
    this.resetActiveTransactions();
    this.resetAccountAssetsSubscription();
    this.resetRuntimeVersionSubscription();
    this.resetFiatPriceAndApySubscription();
  }

  changeTheme(): void {
    switchTheme();
  }
}
</script>

<style lang="scss">
html {
  background: var(--s-color-utility-surface);
}
.el-tooltip__popper.info-tooltip {
  padding: var(--s-basic-spacing);
  max-width: 320px;
  border: none !important;
  box-shadow: var(--s-shadow-tooltip);
  font-size: var(--s-font-size-small);
  line-height: var(--s-line-height-medium);
}
.el-loading-mask {
  background-color: var(--s-color-utility-body);
  .el-loading-spinner {
    background-image: url('~@/assets/img/pswap-loader.svg');
    height: var(--s-size-medium);
    width: var(--s-size-medium);
    margin-left: calc(50% - (var(--s-size-medium) / 2));
    > svg {
      display: none;
    }
  }
}
.el-notification.sora {
  background: var(--s-color-brand-day);
  box-shadow: var(--s-shadow-tooltip);
  border-radius: calc(var(--s-border-radius-mini) / 2);
  border: none;
  align-items: center;
  width: 405px;
  .el-notification {
    &__icon {
      position: relative;
      width: 20px;
      height: 20px;
      border-radius: 50%;
      background: var(--s-color-utility-surface);
      &:before {
        position: absolute;
        top: -2px;
        left: -2px;
      }
    }
    &__content {
      color: var(--s-color-utility-surface);
      text-align: left;
    }
    &__closeBtn {
      color: var(--s-color-utility-surface);
      &:hover {
        color: var(--s-color-utility-surface);
      }
    }
  }
  .loader {
    position: absolute;
    bottom: 0;
    left: 0;
    height: 2px;
    background: var(--s-color-utility-surface);
    // If duration will be change we should create css variable for it
    animation: runloader 4.5s linear infinite;
    @keyframes runloader {
      0% {
        width: 100%;
      }
      100% {
        width: 0;
      }
    }
  }
  &:hover .loader {
    width: 0;
    animation: none;
  }
}
</style>

<style scoped lang="scss">
.wallet-wrapper {
  margin: 40px 0;
  justify-content: center;
  align-items: center;
  height: 100vh;
}
.app-wrapper {
  display: flex;
  justify-content: space-between;
}
.theme-switch,
.hide-balance-switch {
  width: 185px;
  margin: 10px;
}
</style>
