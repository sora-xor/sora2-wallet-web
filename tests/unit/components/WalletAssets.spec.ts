import WalletAssets from '@/components/WalletAssets.vue';

import { useDescribe, useShallowMount, useVuex } from '../../utils';
import {
  MOCK_ACCOUNT_ASSETS,
  MOCK_FIAT_PRICE_AND_APY_OBJECT,
  MOCK_WALLET_PERMISSIONS,
  MOCK_WHITE_LIST,
} from '../../utils/mock';
import { WalletPermissions } from '../../../src/consts';

const createWrapper = (options = {}) =>
  useShallowMount(WalletAssets, {
    ...options,
  });

const createStore = (permissions: WalletPermissions = MOCK_WALLET_PERMISSIONS, withoutFiatAndApy = false) =>
  useVuex({
    settings: {
      state: () => ({
        permissions,
        shouldBalanceBeHidden: false,
        filters: {
          option: 'All',
          verifiedOnly: false,
          zeroBalance: false,
        },
      }),
    },
    account: {
      state: () => ({
        accountAssets: MOCK_ACCOUNT_ASSETS,
        fiatPriceAndApyObject: MOCK_FIAT_PRICE_AND_APY_OBJECT,
        withoutFiatAndApy,
      }),
      getters: {
        whitelist: () => MOCK_WHITE_LIST,
      },
    },
    router: {
      mutations: {
        navigate: jest.fn(),
      },
    },
  });

useDescribe('WalletAssets.vue', WalletAssets, () => {
  it('should be rendered correctly', () => {
    // const wrapper = useShallowMount(WalletAssets, { store: createStore() });

    const wrapper = createWrapper({
      store: createStore(),
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  it('should not render send button when sendAssets property is false', () => {
    const wrapper = createWrapper({
      store: createStore({ ...MOCK_WALLET_PERMISSIONS, sendAssets: false }),
    });
    const sendBtn = wrapper.find('.send');

    expect(sendBtn.exists()).toBeFalse();
  });

  it('should not render swap button when swapAssets property is false', () => {
    const wrapper = createWrapper({
      store: createStore({ ...MOCK_WALLET_PERMISSIONS, swapAssets: false }),
    });
    const swapBtn = wrapper.find('.swap');

    expect(swapBtn.exists()).toBeFalse();
  });

  it('should not render asset details button when showAssetDetails property is false', () => {
    const wrapper = createWrapper({
      store: createStore({ ...MOCK_WALLET_PERMISSIONS, showAssetDetails: false }),
    });
    const detailsBtn = wrapper.find('.el-button--details');

    expect(detailsBtn.exists()).toBeFalse();
  });

  it('should not render add assets button when addAssets property is false', () => {
    const wrapper = createWrapper({
      store: createStore({ ...MOCK_WALLET_PERMISSIONS, addAssets: false }),
    });
    const addAssetsBtn = wrapper.find('.wallet-assets-add');

    expect(addAssetsBtn.exists()).toBeFalse();
  });

  it('should not render fiat value when withoutFiatAndApy property is true', () => {
    const wrapper = createWrapper({
      store: createStore(undefined, true),
    });
    const fiatValue = wrapper.find('.wallet-assets--fiat');

    expect(fiatValue.exists()).toBeFalse();
  });
});
