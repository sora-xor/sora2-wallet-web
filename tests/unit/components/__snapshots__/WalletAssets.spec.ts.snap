// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WalletAssets.vue should be rendered correctly 1`] = `
<div
  class="wallet-assets wallet-assets--fiat s-flex"
>
  <wallet-assets-headline-stub
    assetsfiatamount="151.782"
  />
   
  <s-scrollbar-stub
    class="wallet-assets-scrollbar"
    tag="div"
  >
    <draggable-stub
      class="wallet-assets__draggable"
      clone="[Function]"
      element="div"
      options="[object Object]"
      value="[object Object],[object Object],[object Object]"
    >
      <div
        class="wallet-assets-item__wrapper"
      >
        <div
          class="wallet-assets-item s-flex"
        >
          <div
            class="wallet-assets-dashes"
          >
            <div
              class="wallet-assets-three-dash"
            />
          </div>
           
          <asset-list-item-stub
            asset="[object Object]"
            withclickablelogo="true"
            withfiat="true"
          />
           
          <s-divider-stub
            class="wallet-assets-divider"
            contentposition="center"
            direction="horizontal"
            type="secondary"
          />
        </div>
      </div>
      <div
        class="wallet-assets-item__wrapper"
      >
        <div
          class="wallet-assets-item s-flex"
        >
          <div
            class="wallet-assets-dashes"
          >
            <div
              class="wallet-assets-three-dash"
            />
          </div>
           
          <asset-list-item-stub
            asset="[object Object]"
            withclickablelogo="true"
            withfiat="true"
          />
           
          <s-divider-stub
            class="wallet-assets-divider"
            contentposition="center"
            direction="horizontal"
            type="secondary"
          />
        </div>
      </div>
      <div
        class="wallet-assets-item__wrapper"
      >
        <div
          class="wallet-assets-item s-flex"
        >
          <div
            class="wallet-assets-dashes"
          >
            <div
              class="wallet-assets-three-dash"
            />
          </div>
           
          <asset-list-item-stub
            asset="[object Object]"
            withclickablelogo="true"
            withfiat="true"
          />
           
          <s-divider-stub
            class="wallet-assets-divider"
            contentposition="center"
            direction="horizontal"
            type="secondary"
          />
        </div>
      </div>
       
      <!---->
    </draggable-stub>
  </s-scrollbar-stub>
   
  <s-button-stub
    borderradius="small"
    class="wallet-assets-add s-typography-button--large"
    icon=""
    iconposition="left"
    nativetype="button"
    size="medium"
    tabindex="0"
    tooltip=""
    tooltipplacement="top"
    type="secondary"
    usedesignsystem="true"
  >
    
    Add asset
  
  </s-button-stub>
</div>
`;
